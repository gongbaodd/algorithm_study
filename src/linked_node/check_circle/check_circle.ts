// https://www.cnblogs.com/nipan/p/4098592.html
import { LNode, init } from "../lib";

export function makeLoop(): LNode {
  const head = init([1, 2, 3, 4, 5, 6, 7]);
  let current: LNode = head.next as LNode;

  while (current.next) {
    current = current.next;
  }

  current.next = head.next?.next?.next ?? null;

  return head;
}

export function findMeet(head: LNode): LNode | null {
  let slow = head.next;
  let fast = head.next;

  for (;;) {
    if (fast === null || slow === null) {
      return null;
    }

    slow = slow?.next ?? null;
    fast = fast?.next?.next ?? null;

    if (slow === fast) {
      return slow;
    }
  }
}

export function findEntrance(head: LNode, meetNode: LNode): LNode | null {
  let aCur = head.next;
  let bCur: LNode | null = meetNode;

  while (aCur !== bCur) {
    aCur = aCur?.next ?? null;
    bCur = bCur?.next ?? null;
  }

  return aCur;
}
