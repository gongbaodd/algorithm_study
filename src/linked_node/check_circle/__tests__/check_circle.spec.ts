import { findMeet, makeLoop, findEntrance } from "../check_circle";
import { init, LNode } from "../../lib";

describe("check_circles", () => {
  describe("findMeet", () => {
    describe("normal chain", () => {
      it("should return null", () => {
        const head = init([1, 2, 3, 4, 5]);
        expect(findMeet(head)).toBeNull();
      });
    });

    describe("the chain has loop", () => {
      it("should return the meet point", () => {
        const head = makeLoop();
        expect(findMeet(head)).toMatchSnapshot();
      });
    });
  });

  describe("findEntrance", () => {
    it("should return rntrance node", () => {
      const head = makeLoop();
      const meet = findMeet(head) as LNode;

      expect(findEntrance(head, meet)).toMatchSnapshot();
    });
  });
});
