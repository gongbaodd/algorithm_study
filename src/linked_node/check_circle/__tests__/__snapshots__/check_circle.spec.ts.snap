// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check_circles findEntrance should return rntrance node 1`] = `
LNode {
  "data": 3,
  "next": LNode {
    "data": 4,
    "next": LNode {
      "data": 5,
      "next": LNode {
        "data": 6,
        "next": LNode {
          "data": 7,
          "next": [Circular],
        },
      },
    },
  },
}
`;

exports[`check_circles findMeet the chain has loop should return the meet point 1`] = `
LNode {
  "data": 6,
  "next": LNode {
    "data": 7,
    "next": LNode {
      "data": 3,
      "next": LNode {
        "data": 4,
        "next": LNode {
          "data": 5,
          "next": [Circular],
        },
      },
    },
  },
}
`;
